File: APAC Semifinal 2008/C.cpp
// Problem: Millionaire
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <stdio.h>
#include <iostream>

using namespace std;

const int MAX = 1000000;

double d[16][(1<<15)+1];

double solve()
{
    int m;
    double p, x;
  
    cin >> m >> p >> x;
    
    if (x >= MAX) return 1.0;
    if (x < 1.0/(1<<m)) return 0.0;

    d[0][1] = 1.0;
    for (int j = 1; j <= m; j++)
    {
	d[j][1<<j] = 1.0;
	for (int i = 1; i < (1<<j); i++)
	{
	    double c = 0.0;
	    for (int k = (i+1)/2; k <= (1<<(j-1)); k++)
	    {
		if (i-k >= 0)
		{
		    double t = p*d[j-1][k] + (1-p)*d[j-1][i-k];
		    if (t > c) c = t;
		}
	    }
	    d[j][i] = c;
	}
    }

    return d[m][(int)((1<<m)*(x/MAX))];
    
}

int main()
{
    int cases;
    cin >> cases;

    for (int i = 1; i <= cases; i++)
    {
      printf("Case #%d: %0.6f\n", i, solve());
    }
    
    return 0;
}

File: APAC Semifinal 2008/C.py
#!/usr/bin/env python
#
# Problem: Millionaire
# Language: Python
# Author: KirarinSnow
# Usage: python thisfile.py <input.in >output.out
# Comments: May be too slow for large.


from math import *

MAX = 1000000

def compute():
    m, p, x = raw_input().split()
    m = int(m)
    p = float(p)
    x = int(x)

    if x >= MAX:
        return 1.0

    if x < 1.0/(1<<m):
        return 0.0

    d = [[0.0] * ((1<<m)+1) for i in range(m+1)]
    d[0][1] = 1.0
    for j in range(1, m+1):
        d[j][1] = p**j
        d[j][1<<j] = 1.0
        for i in range(2, (1<<j)):
            c = 0.0
            for k in range((i+1)/2, (1<<(j-1))+1):
                if i-k >= 0:
                    c = max(c, p*d[j-1][k] + (1-p)*d[j-1][i-k])
            d[j][i] = c

    return d[m][int((1<<m)*((x+0.0)/MAX))]
    

for i in range(input()):
    print "Case #%d: %0.6f" % (i+1, compute())
